zebra-state/src/service/check/tests/prop.rs

```diff
@@ -0,0 +1,386 @@
+//! Randomised property tests for state contextual validation nullifier: (), in_finalized_state: ()  nullifier: (), in_finalized_state: ()  checks.
+
+use std::sync::Arc;
+
+use itertools::Itertools;
+use proptest::prelude::*;
+
+use zebra_chain::{
+    block::{Block, Height},
+    fmt::TypeNameToDebug,
+    parameters::Network::*,
+    primitives::Groth16Proof,
+    serialization::ZcashDeserializeInto,
+    sprout::{self, JoinSplit},
+    transaction::{JoinSplitData, LockTime, Transaction},
+};
+
+use crate::{
+    config::Config, service::StateService, tests::Prepare, FinalizedBlock,
+    ValidateContextError::DuplicateSproutNullifier,
+};
+
+// These tests use the `Arbitrary` trait to easily generate complex types,
+// then modify those types to cause an error (or to ensure success).
+//
+// We could use mainnet or testnet blocks in these tests,
+// but the differences shouldn't matter,
+// because we're only interested in spend validation,
+// (and passing various other state checks).
+proptest! {
+    /// Make sure an arbitrary sprout nullifier is accepted by state contextual validation.
+    ///
+    /// This test makes sure there are no spurious rejections that might hide bugs in the other tests.
+    /// (And that the test infrastructure generally works.)
+    #[test]
+    fn accept_distinct_arbitrary_sprout_nullifiers(
+        mut joinsplit in TypeNameToDebug::<JoinSplit::<Groth16Proof>>::arbitrary(),
+        mut joinsplit_data in TypeNameToDebug::<JoinSplitData::<Groth16Proof>>::arbitrary(),
+    ) {
+        zebra_test::init();
+
+        let mut block1 = zebra_test::vectors::BLOCK_MAINNET_1_BYTES
+            .zcash_deserialize_into::<Block>()
+            .expect("block should deserialize");
+
+        make_distinct_nullifiers(&mut joinsplit.nullifiers);
+
+        // make sure there are no other nullifiers
+        joinsplit_data.first = joinsplit.0;
+        joinsplit_data.rest = Vec::new();
+
+        let transaction = transaction_v4_with_joinsplit_data(joinsplit_data.0);
+
+        // convert the coinbase transaction to a version that the non-finalized state will accept
+        block1.transactions[0] = transaction_v4_from_coinbase(&block1.transactions[0]).into();
+
+        block1
+            .transactions
+            .push(transaction.into());
+
+        let (mut state, _genesis) = new_state_with_mainnet_genesis();
+
+        let block1 = Arc::new(block1).prepare();
+        let commit_result =
+            state.validate_and_commit(block1.clone());
+
+        // block was accepted
+        prop_assert_eq!(commit_result, Ok(()));
+        prop_assert_eq!(Some((Height(1), block1.hash)), state.best_tip());
+    }
+
+    /// Make sure duplicate sprout nullifiers are rejected by state contextual validation,
+    /// if they come from the same JoinSplit.
+    #[test]
+    fn reject_duplicate_sprout_nullifiers_in_joinsplit(
```

Comment:
Unfortunately, there are 4 different ways that sprout nullifiers can be duplicated: joinsplit, transaction, block, and chain.